File: codio-intellij/src/main/kotlin/com/wix/codio/ProgressTimer.kt
package com.wix.codio

import java.util.*

abstract class CodioProgressTimerOnUpdateObserver {
    abstract fun run(current: Int, total: Int)
}

abstract class CodioProgressTimerOnFinishObserver {
    abstract fun run()
}

class CodioProgressTimer(val totalLength: Long) {

    companion object {
        private var onFinishObservers = ArrayList<CodioProgressTimerOnFinishObserver>()
        private var onUpdateObservers = ArrayList<CodioProgressTimerOnUpdateObserver>()

        fun onFinish(observer: CodioProgressTimerOnFinishObserver) {
            this.onFinishObservers.add(observer);
        }

        fun onUpdate(observer: CodioProgressTimerOnUpdateObserver) {
            this.onUpdateObservers.add(observer);
        }
    }

    var timer = Timer()
    var timerTask : TimerTask? = null
    var currentSecond: Int = 0



    fun stop() {
        timer.cancel()
    }

    fun run(tutorialTime: Int = 0) {
        try {
            if (this.timerTask != null) { stop() }
            this.currentSecond = tutorialTime;
            this.timerTask = object : TimerTask() {
                override fun run() {
                    currentSecond++
                    if (totalLength > 0 && currentSecond > totalLength / 1000) {
                        onFinishObservers.forEach{ it.run()}
                        onUpdateObservers.forEach{it.run((totalLength/ 1000).toInt(), (totalLength/ 1000).toInt())}
                        stop()
                    } else {
                        onUpdateObservers.forEach{it.run(currentSecond, (totalLength / 1000).toInt())}
                    }
                }
            }
            timer = Timer()
            timer.scheduleAtFixedRate(timerTask, 1000, 1000)

        } catch(e: Exception) {
            e.printStackTrace()
        }
    }
}

File: vscode/src/ProgressTimer.ts
export default class CodioProgressTimer {
  codioLength: number | undefined;
  timer: NodeJS.Timer;
  currentSecond: number;

  private onUpdateObservers: Array<Function> = [];
  private onFinishObservers: Array<Function> = [];

  constructor(codioLengh?: number) {
    this.codioLength = codioLengh;
  }

  onFinish(observer) {
    this.onFinishObservers.push(observer);
  }

  onUpdate(observer) {
    this.onUpdateObservers.push(observer);
  }

  stop() {
    clearInterval(this.timer);
  }

  run(codioTime = 0) {
    try {
      if (this.timer) {
        clearInterval(this.timer);
      }
      this.currentSecond = codioTime;
      this.timer = setInterval(() => {
        this.currentSecond++;
        if (this.codioLength && this.currentSecond > this.codioLength / 1000) {
          this.onFinishObservers.forEach((observer) => observer());
          clearInterval(this.timer);
          this.onUpdateObservers.forEach((observer) => observer(this.codioLength / 1000, this.codioLength / 1000));
        } else {
          this.onUpdateObservers.forEach((observer) => observer(this.currentSecond, this.codioLength / 1000));
        }
      }, 1000);
    } catch (e) {
      console.log('report progress error,', e);
    }
  }
}
